Separates into core and cobspec modules.

saralein opened the pull request.

saralein committed:
Restructure into server and cobspec modules

- Add handling for OPTIONS requests
- Add tea/coffe routes, controller

damonkelley commented on UsingTheServer.md:
```diff
@@ -0,0 +1,157 @@
+        jcenter()
+    }
+    dependencies {
+        classpath 'com.github.jengelman.gradle.plugins:shadow:2.0.1'
```
What does this do?

saralein replied:
I kept reading people having to use plugins to make fat jars in Gradle, so Shadow is a plugin for that.  However, I found another way which doesn't depend on third-party stuff, so I've updated the `build.gradle` and `md`.

damonkelley commented on UsingTheServer.md:
```diff
@@ -0,0 +1,157 @@
+}
+   
+dependencies {
+    compile project(':core')
```
Will a client project always have a `core` project?

saralein replied:
Yeah, it would always have `core` (or `server`) now.

damonkelley replied:
I was under the impression that this would compile a subproject named `server`.

damonkelley commented on UsingTheServer.md:
```diff
@@ -0,0 +1,157 @@
+public class Main {
+    public static void main(String[] args) {
+        String home = System.getProperty("user.dir");
+        com.saralein.core.Main.main(args, home, Main::setupRoutes);
```
Seems like you may want to not use `Main`. Perhaps `Server` or something?

saralein replied:
Updated it `Main` in `server` to `HttpServer`.

damonkelley reviewed (commented):
Why did you decide to go with `com.saralein.core` instead of `com.saralein.server`?

saralein commented:
Initially I had the modules organized inside another module named `com.saralein.server`, so I couldn't name the internal one `com.saralein.server`.  I've updated it now.

zdavy commented on UsingTheServer.md:
```diff
@@ -0,0 +1,155 @@
+httpServer    
+|_ cobspec      
+|_ server   
+|_ YOUR_APPLICATION
```
seems weird to have to clone this project and build in the directory and modify the existing code, I'd rather have controller over my app without worrying about breaking anything. Is there anyway I can just use my own Main, and if I'm using gradle add:

```
dependencies {
    compile group: 'com.saralein', name: 'server', version: '{latest}'
}
```
?

zdavy commented on UsingTheServer.md:
```diff
@@ -0,0 +1,130 @@
+
+Use of the `ResponseBuilder` is not required but is provided for convenience.
+
+Your `Controller` may contain whatever additional methods are needed to create contents of your response.
```
The public API is well documented, but I still don't understand how to set up and use the server in my own project?, maybe a simple example of wiring it all up would help?

damonkelley replied:
+1

zdavy commented on cobspec/src/main/java/com/saralein/cobspec/Main.java:
```diff
@@ -0,0 +1,78 @@
+                    .addRoute("/coffee", Methods.GET.name(), new CoffeeController())
+                    .build();
+    }
+}
```
Whoa, there seems to be a lot of domain knowledge required to configure and use this server. Is there a way to give it a much simpler interface? It might encourage more people to use if it's easy to use. Just spitballing here, so I might have the syntax way wrong but what if it were like

```java
public static void Main(String[] args) {
  Logger logger = new Logger(new ConnectionLogger(System.Out));
  new Server(logger).configure((RoutesBuilder router) -> {
   return router
      .addStatic("/public");
      .get("/redirect", () -> new RedirectController());
      // ....
  }).start(args);
}
```

damonkelley commented on cobspec/src/main/java/com/saralein/cobspec/Main.java:
```diff
@@ -0,0 +1,78 @@
+            Routes routes = setupRoutes(root);
+            HttpServer.start(port, root, routes, logger);
+        } else {
+            logger.log(String.join("\n", validationErrors));
```
This is a bit minor, but it looks like if the validation failed, you would have an exit code of 0, which implies that the process completed successfully.

saralein commented:
I made a few changes on this branch, but a lot of the effort to fix setup is now in [this PR](https://github.com/saralein/java-server/pull/17).

damonkelley commented on server/src/main/java/com/saralein/server/HttpServer.java:
```diff
@@ -0,0 +1,45 @@
+import com.saralein.server.response.ResponseSerializer;
+import com.saralein.server.router.Routes;
+import com.saralein.server.router.Router;
+import com.saralein.server.setup.Server;
```
Why is `Server` in the `com.saralein.server.setup` package. It feels like it would belong in `com.saralein.server`

damonkelley reviewed (commented):
I am not sure I fully understand the difference between `server.setup.SetupServer`, `server.setup.Server`, and `server.HttpServer`? Can you explain?

saralein commented:
The naming was a mess.  I've tried renaming them a bit.  `Server` is still `Server` because I feel like that piece is truly the server (moved to `com.saralein.server`).  I never thought `HttpServer` was appropriate.  I've renamed it `Application`, since it feels like configuring the application (maybe more so on the basic auth branch).  I renamed `SetupServer` to `ServerInitializer`, which hopefully gives a sense that it takes stuff in, then initializes and returns the server.  I've found these three pieces hard to name.  Or rather, two of them hard to name in relation to `Server`.

damonkelley reviewed (approved):
Ah! Those names make more sense to me.
